cmake-modules/TSPCompile.cmake

```diff
@@ -12,31 +12,96 @@ macro(DownloadTSPFiles TSP_REPO TSP_SHA TSP_REPO_PATH TSP_DESTINATION)
         message(FATAL_ERROR "Git not found")
     endif()
     
-    set(DOWNLOAD_FOLDER ${CMAKE_SOURCE_DIR}/build/${TSP_DESTINATION})
-    #cleanup folder
-    file(REMOVE_RECURSE ${DOWNLOAD_FOLDER})
-    make_directory(${DOWNLOAD_FOLDER})
+    set(TSP_REPO "https://github.com/Azure/azure-rest-api-specs.git")
+    set(DOWNLOAD_TSP_FOLDER ${CMAKE_SOURCE_DIR}/build/${TSP_DESTINATION})
+    # if we have the git folder, we don't need to download it again
+    # this also saves times on incremental builds
+    if(NOT EXISTS ${DOWNLOAD_TSP_FOLDER}/.git)
+    message("First time setting up the ${TSP_DESTINATION} repo.")
+        #make folder
+        make_directory(${DOWNLOAD_TSP_FOLDER})
 
     #init git in folder
     execute_process(COMMAND ${GIT_EXECUTABLE} init
-        WORKING_DIRECTORY ${DOWNLOAD_FOLDER})
+            WORKING_DIRECTORY ${DOWNLOAD_TSP_FOLDER})
     #add remote
     execute_process(COMMAND ${GIT_EXECUTABLE} remote add origin ${TSP_REPO}
-         WORKING_DIRECTORY ${DOWNLOAD_FOLDER})
+             WORKING_DIRECTORY ${DOWNLOAD_TSP_FOLDER})
     #set sparse-checkout
     execute_process(COMMAND ${GIT_EXECUTABLE} sparse-checkout init --cone
-         WORKING_DIRECTORY ${DOWNLOAD_FOLDER})
+             WORKING_DIRECTORY ${DOWNLOAD_TSP_FOLDER})
     #set sparse-checkout folder
     execute_process(COMMAND ${GIT_EXECUTABLE} sparse-checkout set ${TSP_REPO_PATH}
-         WORKING_DIRECTORY ${DOWNLOAD_FOLDER})
+             WORKING_DIRECTORY ${DOWNLOAD_TSP_FOLDER})
+    else()
+        message("Repo detected at ${TSP_DESTINATION}. Jumping ahead to checkout.") 
+    endif()
     #fetch
     execute_process(COMMAND ${GIT_EXECUTABLE} fetch
-         WORKING_DIRECTORY ${DOWNLOAD_FOLDER})
+         WORKING_DIRECTORY ${DOWNLOAD_TSP_FOLDER})
     #switch branch
     execute_process(COMMAND ${GIT_EXECUTABLE} checkout ${TSP_SHA}
-          WORKING_DIRECTORY ${DOWNLOAD_FOLDER})
+          WORKING_DIRECTORY ${DOWNLOAD_TSP_FOLDER})
  
-    if (NOT ${STATUS_CODE} EQUAL 0)
-        message(FATAL_ERROR "TSP download failed (Link: ${TSP_FULL_PATH}).")
+    if(NOT ${STATUS_CODE} EQUAL 0)
+        message(FATAL_ERROR "TSP download failed.")
+    endif()
+endmacro()
+
+macro (DownloadCodeGenerator CODEGEN_SHA CODEGEN_DESTINATION)
+    message("Downloading CODEGEN files using the following params  CODEGEN_REPO=${CODEGEN_REPO} CODEGEN_SHA=${CODEGEN_SHA} CODEGEN_DESTINATION=${CODEGEN_DESTINATION}")
```

Comment:
How this function works is fine for our initial bootstrapping.

Please add a TODO linked to an issue:
  "Download the typspec-cpp emitter from npmjs.org"